src/statsmodels/formula.jl

```diff
@@ -444,59 +455,79 @@ creating the model matrix.
 @compat function (::Type{ModelMatrix{T}}){T<:AbstractFloatMatrix}(mf::ModelFrame)
     dfrm = mf.df
     terms = droprandomeffects(dropresponse!(mf.terms))
+    factors = terms.factors
+
+    ## Step 1:
+    ## compute which columns correspond to which terms in order to pre-allocate
+    ## model matrix and set up views later
+    term_col_inds = []
+    num_cols = convert(Int, terms.intercept)
+
+    for (i_term, term) in enumerate(terms.terms)
+        f = Compat.view(factors, :, i_term)
+        any(f) || continue
+
+        block_sizes = map((et,nr) -> nc(et, mf, non_redundant=nr),
+                          Compat.view(terms.eterms, f),
+                          Compat.view(terms.is_non_redundant, f, i_term))
+        num_term_cols = reduce(*, block_sizes)
```

Comment:
Nope. I think I first tried to write this as a `mapreduce` but `prod` makes more sense now that you mention it.